dask/order.py

```diff
@@ -319,35 +315,77 @@ def finish_now_key(x):
+                    # inner stacks (and in `already_seen`).
+                    #
+                    # Hence, this can introduce locally sub-optimal behavior (more memory, and more
+                    # work that may not go towards our tactical goal) with the goal of being more
+                    # globally optimal.  The risk from being locally sub-optimal is most likely low,
+                    # and the reward for being more globally optimal is potentially very high, so I
+                    # think this strategy is probably a good idea in general.  Heh, even so, I expect
+                    # there are pathological failure cases to be found :)
+                    result[item] = i
+                    i += 1
+                    if metrics[item][3] == 1:  # min_height
+                        # Don't leave any dangling single nodes!  Finish all dependents that are
+                        # ready and are also root nodes.
+                        finish_now = {
+                            dep
+                            for dep in deps
+                            if not dependents[dep] and num_needed[dep] == 1
+                        }
+                        if finish_now:
+                            deps -= finish_now  # Safe to mutate
+                            if len(finish_now) > 1:
+                                finish_now = sorted(finish_now, key=finish_now_key)
+                            for dep in finish_now:
+                                result[dep] = i
+                                i += 1
+                            add_to_inner_stack = False
+                    if deps:
+                        for dep in deps:
+                            num_needed[dep] -= 1
+
+                        already_seen = deps & seen
+                        if already_seen:
+                            if len(deps) == len(already_seen):
+                                deps = None
+                                break
+                            add_to_inner_stack = False
+                            deps -= already_seen
+
+                        if len(deps) == 1:
+                            # Fast path!  We trim down `deps` above hoping to reach here.
+                            (item,) = deps
+                            if num_needed[item] == 0:
+                                deps = dependents[item]
+                                key = partition_keys[item]
+                                if key < item_key:
+                                    # We have three reasonable choices for item_key: first, last, min.
+                                    # I don't have a principled reason for choosing any of the options.
+                                    # So, let's choose min.  It may be the most conservative by avoiding
+                                    # accidentally going down an expensive-to-compute path.
+                                    item_key = key
+                                continue
```

Comment:
FYI, if we remove these lines, then the code logic is exactly equivalent to what it was before.